libsolidity/analysis/TypeChecker.cpp

```diff
@@ -1729,31 +1730,80 @@ bool TypeChecker::visit(UnaryOperation const& _operation)
 		requireLValue(_operation.subExpression(), false);
 	else
 		_operation.subExpression().accept(*this);
-	Type const* subExprType = type(_operation.subExpression());
-	TypeResult result = type(_operation.subExpression())->unaryOperatorResult(op);
-	if (!result)
+	Type const* operandType = type(_operation.subExpression());
+
+	// Check if the operator is built-in or user-defined.
+	TypeResult builtinResult = operandType->unaryOperatorResult(op);
+	set<FunctionDefinition const*, ASTNode::CompareByID> matchingDefinitions = operandType->operatorDefinitions(
+		op,
+		*currentDefinitionScope(),
+		true // _unary
+	);
+
+	// Operator can't be both user-defined and built-in at the same time.
+	solAssert(!builtinResult || matchingDefinitions.empty());
+
+	// By default use the type we'd expect from correct code. This way we can continue analysis
+	// of other expressions in a sensible way in case of a non-fatal error.
+	Type const* resultType = operandType;
```

Comment:
Well, I guess I was thinking mostly about the user-defined case here and there the result type has to match the operand type, right?

Do you mean that built-in stuff like, e.g. `!` returns `bool` rather than operand type?

Reply:
Ah, sorry, actually for the currently allowed unary ops, it's true - and for binary ones below it's actually more specific.

Reply:
We just need to be careful when extending this for more operators for which this may no longer be the case :-).

Reply:
I mean, I should probably reword this to account for `!`. I did not actually think about this particular example back then.